### ISSUE TITLE
UnicodeError Gunicorn generate_latest multiprocess

### ISSUE THREAD

**Issue by saley89** [prometheus/client_python#165]
Hi all,

We are currently seeing the following error from our prometheus endpoint:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/site-packages/cats/server/app_log.py", line 37, in _apply_logging
    result = fn(*args, **kwargs)
  File "/usr/local/lib/python3.4/site-packages/cats/server/prometheus_views.py", line 14, in prometheus_view
    output = generate_latest(registry)
  File "/usr/local/lib/python3.4/site-packages/prometheus_client/exposition.py", line 63, in generate_latest
    for metric in registry.collect():
  File "/usr/local/lib/python3.4/site-packages/prometheus_client/core.py", line 97, in collect
    for metric in collector.collect():
  File "/usr/local/lib/python3.4/site-packages/prometheus_client/multiprocess.py", line 24, in collect
    d = core._MmapedDict(f)
  File "/usr/local/lib/python3.4/site-packages/prometheus_client/core.py", line 335, in __init__
    for key, _, pos in self._read_all_values():
  File "/usr/local/lib/python3.4/site-packages/prometheus_client/core.py", line 365, in _read_all_values
    yield encoded.decode('utf-8'), value, pos
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 102: invalid start byte
```

Weirdly, at first the endpoint responds with our metrics as expected, then after a short period of time as requests come in we start to get the above.  We are finding it impossible to debug what character is causing the issue.  

We only have two metrics:

```
METRICS['application_latency'] = Histogram('application_latency',
                                           "Metrics for application latency",
                                           ['resourceName', 'method', 'worker'])



METRICS['application_responses'] = Counter(
    'application_responses',
    'Metrics for application responses per endpoint, method and http response code',
    ['resourceName', 'method', 'status', 'worker'])
```

And increment them like so:

```
        METRICS['application_responses'].labels(resourceName=resource_name, method=method, status=status_code,
                                                worker=str(os.getpid())).inc()

        METRICS['application_latency'].labels(resourceName=resource_name, method=method,
                                              worker=str(os.getpid())).observe(response_time)
```

All the labels are just strings and the most complex is resourceName which looks like "service.my_method" for example.  Any pointers on how we could possibly diagnose this would be great as currently we are completely stuck.  
We have followed the gunicorn multiprocess example exactly.

Our gunicorn conf is:

```
PROMETHEUS_METRICS = '/prometheus-metrics'

bind = '0.0.0.0:9077'
worker_class = 'gevent'
workers = 8
worker_connections = 8000
worker_tmp_dir = '/worker-tmp'
backlog = 16384
keepalive = 8


def on_starting(_server):
    [os.remove("{}/{}".format(PROMETHEUS_METRICS, f)) for f in os.listdir(PROMETHEUS_METRICS)]

def child_exit(_server, worker):
    from prometheus_client import multiprocess
    multiprocess.mark_process_dead(worker.pid)
```

We also set this environment variable before starting:

```
prometheus_multiproc_dir=/prometheus-metrics
```

Finally, we are running kubernetes hosted on AWS.

**Comment by brian-brazil** [comment]
I'm guessing you've messed up bytes vs unicode in one of your label values.

You also don't need the `worker` label, the multiprocessing code takes care of that already.

**Comment by saley89** [comment]
Thanks for the quick reply @brian-brazil.  Should we be using bytes for the label values then?

Also, without the worker label, we get load balanced between the workers whenever we hit the prometheus metrics endpoint, which we assumed would cause the metrics to appear wrongly in prometheus and grafana. 

Can you clarify what you mean by it taking care of it?

**Comment by brian-brazil** [comment]
Label values should be utf-8.

> without the worker label, we get load balanced between the workers whenever we hit the prometheus metrics endpoint

You mustn't have setup the multi process mode correctly.

**Comment by saley89** [comment]
So we should encode any label values, e.g

```
        METRICS['application_latency'].labels(resourceName=resource_name.encode('utf-8'), method=method.encode('utf-8'),
                                              worker=str(os.getpid()).encode('utf-8')).observe(response_time)
```

I'll review the multiprocess setup.  Could be something around how we setup our environment variable.

**Comment by saley89** [comment]
So i've just reviewed the multiprocess setup.  Ive removed the worker label and what I find is that the metrics increment randomly.  As if some of the requests are lost.  So we aren't being load balanced across workers its just missing some of the requests.

Do you have any pointers for what I can check to ensure its setup correctly.  

We have the environment variable. It's pointing to a directory that it can write too.  
We have the Gunicorn hook.
And this is our prometheus view which is called every time I hit `/private/prometheus`:

```
from prometheus_client import CollectorRegistry, multiprocess, generate_latest, CONTENT_TYPE_LATEST

@prometheus.route('/private/prometheus/')
@apply_logging
def prometheus_view():
    registry = CollectorRegistry()
    multiprocess.MultiProcessCollector(registry)
    output = generate_latest(registry)
    return Response(output, mimetype=CONTENT_TYPE_LATEST)
```

**Comment by brian-brazil** [comment]
That all looks correct. Are the files getting created?

**Comment by saley89** [comment]
They are indeed.  I get the following created in my /prometheus-metrics directory:

```
ls -la /prometheus-metrics/
total 4096
drwxrwxrwx   4 USERNAME  staff   136B  1 Jun 10:38 ./
drwxr-xr-x  36 root    wheel   1.3K 31 May 16:09 ../
-rw-r--r--   1 USERNAME  staff   1.0M  1 Jun 10:38 counter_73606.db
-rw-r--r--   1 USERNAME  staff   1.0M  1 Jun 10:38 histogram_73606.db
```

**Comment by brian-brazil** [comment]
That's only for one process.

**Comment by saley89** [comment]
So I have 8 workers, so your saying there should be 8 (well 16 for two metrics) files?

Any ideas why this would be the case?  I presume that number on the end is a pid? I checked using ps aux and there was nothing for that pid and it didnt match any gunicorn process pid.

**Comment by saley89** [comment]
Ok you got it.  We had some code doing initialising of metrics in creation of the app. 

I've moved some things around and now I see .db files for all 8 workers.

In terms of the original issue I have wrapped all label values with a call to this:

```
def _decode(value):
    try:
        str(value, 'utf-8')
    except TypeError:
        return str(value)
```

Which forces anything that's bytes into 'utf-8' or just str for other types like int etc.  So far I haven't seen a re-occurence of the issue.   

I'd like to add thanks for the replies and help so far. Its starting to make sense now!
